src/standard/expr.mli

```diff
@@ -16,736 +16,99 @@ type hash = private int
 type index = private int
 type 'a tag = 'a Tag.t
 
-type let_kind =
-  | Parallel
-  | Sequential
+(** {3 Type definitions} *)
 
-type builtin = ..
-(* Extensible variant type for builtin operations. Encodes in its type
-   arguments the lengths of the expected ty and term arguments respectively. *)
+type builtin = < ty_cst : ty_cst ; term_cst : term_cst > Builtin.t
+(* Extensible variant type for builtin operations. *)
```

Comment:
Ok, so about the builtins:
- for the upcoming work on printing terms, it was necessary for the builtins to be in a separate module, and to be parameterized (since the printers will also be parameterized/functorized). 
- however, some of the builtins (mainly the ones about ADT constructors/destructors/testers) carry in their payload some terms, and thus, the builtin type had to be parameterized over all that could appear within.
- in a first try I had explicit type parameters for builtins for each potential type/term construction, but that meant having some 6 type parameters, which was... annoying to use let's say
- thus I used the trick of using a phantom object type as some kind of "type dictionary" of type parameters (each method correspond to a type parameter), which ends up being quite nice, since it brings subtyping of builtins as needed, and can easily be extended.
Though, indeed, now that I re-read this, to allow users to correctly extend builtins, I'll need to add all the usable type parameters in the definition in `expr.ml/expr.mli`, i'll do that asap.